sdk/core/core/inc/az_str.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_STR_H
+#define AZ_STR_H
+
+#include <az_assert.h>
+
+#include <az_static_assert.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <limits.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// An immutable string slice.
+typedef struct {
+  // Points to the first byte (character).
+  uint8_t const *begin;
+  // A numer of octets (bytes) in the string.
+  size_t size;
+} az_const_str;
+
+// A muttable string slice.
+typedef struct {
+  uint8_t *begin;
+  size_t size;
+} az_str;
+
+AZ_STATIC_ASSERT(CHAR_BIT == 8);
+
+// A size of the string literal.
+// Details: to make sure that `S` is a `string literal`, we are appending `""` to `S`.
+#define AZ_STRING_LITERAL_LEN(S) (sizeof(S "") - 1)
+
+#define AZ_CONST_STR_DECL(NAME, STRING_LITERAL) \
+  static az_const_str const NAME = { .begin = STRING_LITERAL, .size = AZ_STRING_LITERAL_LEN(STRING_LITERAL) };
+
+#define AZ_CONST_STR(STRING_LITERAL) \
+  (az_const_str){ .begin = STRING_LITERAL, .size = AZ_STRING_LITERAL_LEN(STRING_LITERAL) }
+
+inline uint8_t const *az_const_str_ptr(az_const_str const buffer, size_t const index) {
+  AZ_ASSERT(index <= buffer.size);
+  return buffer.begin + index;
+}
+
+inline uint8_t az_const_str_item(az_const_str const buffer, size_t const index) {
+  AZ_ASSERT(index < buffer.size);
+  return buffer.begin[index];
+}
+
+inline uint8_t const *az_const_str_end(az_const_str const buffer) {
```

Comment:
This returns a pointer to the end of a slice? If so, I don't know why we'd ever need this. I shouldn't ever need to get to the end of a current slice.

Reply:
Initially, my implimentation was more like C++ with begin/end iterators but then I changed it to using indeces. You are right, we don't need this function any more.